Firmware/MVP/battsci.cpp

```diff
@@ -86,6 +98,169 @@ uint8_t BATTSCI_calculateChecksum( uint8_t frameSum )
+  } else if (LiBCM_SoC >= 5) {
+    calculatedSoC = map(LiBCM_SoC, 5, 29, 1, 150);      // BG Regen Allowed, SoC very low.
+  } else {
+    // No Assist Allowed
+    calculatedSoC = 0;
+    oldCalculatedSoC = 0;         // Make sure SoC doesn't immediately spike back up
+    BATTSCI_evaluateSoCBytes(calculatedSoC);
+  }
+
+  // On first run we set oldCalculatedSoC and calculatedSoC to the same value.
+  if (initializeSoC) {
+    oldCalculatedSoC = calculatedSoC;
+    initializeSoC = false;
+    BATTSCI_evaluateSoCBytes(calculatedSoC);
+  }
+
+  static int16_t packMilliAmps = 0;
+  packMilliAmps = adc_getLatestBatteryCurrent_amps();
+
+  // Modify calculatedSoC in place to be an increment of oldCalculatedSoC
+  // packMilliAmps is being checked so that we only increment SoC if we have current < -0.01 Amps or decremented if we have > +0.01 Amps
+  // This should prevent SoC being changed during Auto Stop.
+
+  if (calculatedSoC > oldCalculatedSoC) {
+	if (packMilliAmps <= -300) {								// packMilliAmps is - if Amps are going IN and voltage is going UP
+		if ((oldCalculatedSoC + 5) > calculatedSoC) {
+			calculatedSoC = (oldCalculatedSoC + 5);
+			BATTSCI_evaluateSoCBytes(calculatedSoC);
+			oldCalculatedSoC = calculatedSoC;
+		} else if ((oldCalculatedSoC + 2) > calculatedSoC) {
+			calculatedSoC = (oldCalculatedSoC + 2);
+			BATTSCI_evaluateSoCBytes(calculatedSoC);
+			oldCalculatedSoC = calculatedSoC;
+		} else if ((oldCalculatedSoC + 1) > calculatedSoC) {
+			calculatedSoC = (oldCalculatedSoC + 1);
+			BATTSCI_evaluateSoCBytes(calculatedSoC);
+			oldCalculatedSoC = calculatedSoC;
+		}
+	} else if (packMilliAmps <= -10 ) {
+      calculatedSoC = (oldCalculatedSoC + 1);
+      BATTSCI_evaluateSoCBytes(calculatedSoC);
+      oldCalculatedSoC = calculatedSoC;
+    }
+  } else if (calculatedSoC < oldCalculatedSoC) {
+	if (packMilliAmps >= 300 ) {								// packMilliAmps is + if Amps are going OUT and voltage is going DOWN
+		if ((oldCalculatedSoC - 5) < calculatedSoC) {
+			calculatedSoC = (oldCalculatedSoC - 5);
+			BATTSCI_evaluateSoCBytes(calculatedSoC);
+			oldCalculatedSoC = calculatedSoC;
+		} else if ((oldCalculatedSoC - 2) < calculatedSoC) {
+			calculatedSoC = (oldCalculatedSoC - 2);
+			BATTSCI_evaluateSoCBytes(calculatedSoC);
+			oldCalculatedSoC = calculatedSoC;
+		} else if ((oldCalculatedSoC - 1) < calculatedSoC) {
+			calculatedSoC = (oldCalculatedSoC - 1);
+			BATTSCI_evaluateSoCBytes(calculatedSoC);
+			oldCalculatedSoC = calculatedSoC;
+		}
+    } else if (packMilliAmps >= 10 ) {
+      calculatedSoC = (oldCalculatedSoC - 1);
+      BATTSCI_evaluateSoCBytes(calculatedSoC);
+      oldCalculatedSoC = calculatedSoC;
+    }
+  }
+
+  BATTSCI_evaluateTemperatureByte(calculatedSoC);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void BATTSCI_initializeMCMSoC(void)
+{
+  // Key On causes a recalculation of SoC to spoof to MCM
+	initializeSoC = true;
```

Comment:
Most of the work I'm doing is inside BATTSCI_calculateSoC()

BATTSCI_calculateSoC() doesn't ever run while the car is turned off because it's only triggered with every 17th 0x87 frame sent to the MCM.

I'm using oldCalculatedSoC and calculatedSoC to blend the spoofed SoC values sent to the MCM.  I need those values to get cleared and set to align with whatever SoC_getBatteryStateNow_percent() says it is immediately at key on, instead of blending towards that value from wherever the spoofed SoC was the last time the car got turned off.  This matters most if you grid charge the car.